File: Java/PrisonerSimulator/src/Game/State/ProcessGameState.java
package Game.State;

import Game.Decision;
import Game.Game;
import Player.IPlayerPlayFacade;

import java.util.ArrayList;
import java.util.List;

public abstract class ProcessGameState implements IGameState {
    protected final Game game;

    protected List<IPlayerPlayFacade> currentCompetitors;

    protected int currentPlayerIndex;

    protected int currentRivalIndex;

    public ProcessGameState(Game game) {
        this.game = game;
        currentCompetitors = new ArrayList<>();
        currentPlayerIndex = 0;
        currentRivalIndex = 1;
    }

    /**
     * This method assumes that indexes are correct and does not update them after the step
     */
    protected void RunCompetitionStep() {
        var player = currentCompetitors.get(currentPlayerIndex);
        var rival = currentCompetitors.get(currentRivalIndex);

        var playerId = player.getId();
        var rivalId = rival.getId();

        var playerDecision = player.makeDecision(rivalId);
        var rivalDecision = rival.makeDecision(rivalId);

        player.storeDecision(rivalId, rivalDecision);
        rival.storeDecision(playerId, playerDecision);

        var gameRules = game.getGameRules();

        if (playerDecision == Decision.Cooperate && rivalDecision == Decision.Cooperate) {
            player.updateScore(gameRules.getCooperationReward(), rivalId);
            rival.updateScore(gameRules.getCooperationReward(), playerId);
            return;
        }

        if (playerDecision == Decision.Defect && rivalDecision == Decision.Defect) {
            player.updateScore(gameRules.getDefectFine(), rivalId);
            rival.updateScore(gameRules.getDefectFine(), playerId);
            return;
        }

        if (playerDecision == Decision.Cooperate && rivalDecision == Decision.Defect) {
            player.updateScore(gameRules.getCooperationFine(), rivalId);
            rival.updateScore(gameRules.getDefectReward(), playerId);
            return;
        }

        if (playerDecision == Decision.Defect && rivalDecision == Decision.Cooperate) {
            player.updateScore(gameRules.getDefectReward(), rivalId);
            rival.updateScore(gameRules.getCooperationFine(), playerId);
        }
    }
}

File: Kotlin/PrisonerSimulator/src/game/state/ProcessGameState.kt
package game.state

import game.Configurator
import game.Decision.Cooperate
import game.Decision.Defect
import game.Game
import player.IPlayerPlayFacade

abstract class ProcessGameState(protected val game: Game) : IGameState {
    protected abstract val currentCompetitors: List<IPlayerPlayFacade>
    protected var currentPlayerIndex = 0
    protected var currentRivalIndex = 1

    // This method assumes that indexes are correct and does not update them after the step
    protected open fun runCompetitionStep() {
        val (player, rival) = currentCompetitors[currentPlayerIndex] to currentCompetitors[currentRivalIndex]

        val (playerDecision, rivalDecision) = player.makeDecision(rival.id) to rival.makeDecision(player.id)

        player.storeDecision(rival.id, rivalDecision)
        rival.storeDecision(player.id, playerDecision)

        val (playerDelta, rivalDelta) = when {
            playerDecision == Cooperate && rivalDecision == Cooperate ->
                Configurator.cooperationReward to Configurator.cooperationReward

            playerDecision == Defect && rivalDecision == Defect ->
                Configurator.defectFine to Configurator.defectFine

            playerDecision == Cooperate && rivalDecision == Defect ->
                Configurator.cooperationFine to Configurator.defectReward

            playerDecision == Defect && rivalDecision == Cooperate ->
                Configurator.defectReward to Configurator.cooperationFine

            else -> throw IllegalStateException("Unknown decision: $playerDecision or $rivalDecision")
        }

        player.updateScore(playerDelta, rival.id)
        rival.updateScore(rivalDelta, player.id)
    }
}